src/arch/src/aarch64/mod.rs

```diff
@@ -43,13 +46,23 @@ pub fn arch_memory_regions(
     initrd_size: u64,
     firmware_size: Option<usize>,
 ) -> (ArchMemoryInfo, Vec<(GuestAddress, usize)>) {
+    let ram_start_addr = if firmware_size.is_some() {
+        DRAM_MEM_START_EFI
+    } else {
+        DRAM_MEM_START_KERNEL
+    };
     let page_size: usize = unsafe { libc::sysconf(libc::_SC_PAGESIZE).try_into().unwrap() };
     let dram_size = align_upwards!(size, page_size);
-    let ram_last_addr = layout::DRAM_MEM_START + (dram_size as u64);
+    let ram_last_addr = ram_start_addr + (dram_size as u64);
     let shm_start_addr = ((ram_last_addr / 0x4000_0000) + 1) * 0x4000_0000;
-    let fdt_addr = ram_last_addr - layout::FDT_MAX_SIZE as u64;
+    let fdt_addr = if firmware_size.is_some() {
+        DRAM_MEM_START_EFI
```

Comment:
So if EFI firmware exists, the FDT is placed at the _beginning_ of DRAM.
```
let fdt_addr = if firmware_size.is_some() {
        DRAM_MEM_START_EFI
```

Yet, if direct kernel boot, the FDT is placed at the _end_ of DRAM?
```
} else {
        ram_last_addr - layout::FDT_MAX_SIZE as u64
};
```

Reply:
Yes. For direct kernel boot on aarch64, the kernel *must* be placed at 0x8000_0000. Ideally, DRAM should also start there, because otherwise it needs to be large enough to cover both the start address and 0x8000_0000 (this is what this PR is all about). This means we can't put FDT at the beginning of DRAM, so we do the next best thing, which is putting it at the end.

On the other hand, EDK2 is firmware so it's loaded outside DRAM (at 0x0, while DRAM starts at 0x4000_0000), and expects to find the FDT at 0x4000_0000. So we're pretty much stuck with having to do this different memory layout dance.